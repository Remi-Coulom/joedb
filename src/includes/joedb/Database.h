#ifndef joedb_Database_declared
#define joedb_Database_declared

#include "Readable_Writeable.h"
#include "Table.h"

#include <map>

namespace joedb
{
 class Database: public Readable_Writeable
 {
  private:
   const record_id_t max_record_id;
   table_id_t current_table_id = 0;
   std::map<table_id_t, Table> tables;

  public:
   Database(record_id_t max_record_id = 0): max_record_id(max_record_id) {}

   //
   // Readable override
   //
   record_id_t get_max_record_id() const override {return max_record_id;}
   const std::map<table_id_t, Table> &get_tables() const override
   {
    return tables;
   }
   size_t get_current_table_id() const override {return current_table_id;}
   table_id_t find_table(const std::string &name) const override;
   field_id_t find_field(table_id_t table_id,
                         const std::string &name) const override;
   Type::type_id_t get_field_type(table_id_t table_id,
                                  field_id_t field_id) const override;

   //
   // Writeable override
   //
   void create_table(const std::string &name) override;
   void drop_table(table_id_t table_id) override;
   void rename_table(table_id_t table_id, const std::string &name) override;
   void add_field(table_id_t table_id,
                  const std::string &name,
                  Type type) override;
   void drop_field(table_id_t table_id, field_id_t field_id) override;
   void rename_field(table_id_t, field_id_t, const std::string &name) override;

   void custom(const std::string &name) override {};
   void comment(const std::string &comment) override {};
   void timestamp(int64_t timestamp) override {};
   void valid_data() override {};

   void insert_into(table_id_t table_id, record_id_t record_id) override;
   void insert_vector(table_id_t table_id,
                      record_id_t record_id,
                      record_id_t size) override;
   void delete_from(table_id_t table_id, record_id_t record_id) override;

   #define TYPE_MACRO(type, return_type, type_id, R, W)\
   void update_##type_id(table_id_t table_id,\
                         record_id_t record_id,\
                         field_id_t field_id,\
                         return_type value) override;\
   void update_vector_##type_id(table_id_t table_id,\
                                record_id_t record_id,\
                                field_id_t field_id,\
                                record_id_t size,\
                                const type *value) override;
   #include "TYPE_MACRO.h"
   #undef TYPE_MACRO
 };
}

#endif
