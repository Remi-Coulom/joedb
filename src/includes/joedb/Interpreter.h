#ifndef joedb_Interpreter_declared
#define joedb_Interpreter_declared

#include <iosfwd>

#include "joedb/Type.h"

namespace joedb
{
 class Readable_Writeable;

 class Interpreter
 {
  private:
   Readable_Writeable &db;

   Type parse_type(std::istream &in, std::ostream &out);
   Table_Id parse_table(std::istream &in, std::ostream &out);
   void update_value(std::istream &in,
                     Table_Id table_id,
                     Record_Id record_id,
                     Field_Id field_id);

  public:
   Interpreter(Readable_Writeable &db): db(db) {}

   void main_loop(std::istream &in, std::ostream &out);
 };
}

#endif
