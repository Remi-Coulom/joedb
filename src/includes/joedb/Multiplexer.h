#ifndef joedb_Multiplexer_declared
#define joedb_Multiplexer_declared

#include <vector>

#include "Writeable.h"

namespace joedb
{
 class Multiplexer: public virtual Writeable
 {
  private:
   std::vector<Writeable *> writeables;

  public:
   void add_writeable(Writeable &writeable);

   void create_table(const std::string &name) override;
   void drop_table(table_id_t table_id) override;
   void rename_table(table_id_t table_id, const std::string &name) override;
   void add_field(table_id_t table_id,
                  const std::string &name,
                  Type type) override;
   void drop_field(table_id_t table_id, field_id_t field_id) override;
   void rename_field(table_id_t table_id,
                     field_id_t field_id,
                     const std::string &name) override;
   void custom(const std::string &name) override;
   void comment(const std::string &comment) override;
   void timestamp(int64_t timestamp) override;
   void valid_data() override;
   void insert(table_id_t table_id, record_id_t record_id) override;
   void insert_vector(table_id_t table_id,
                      record_id_t record_id,
                      record_id_t size) override;
   void delete_record(table_id_t table_id, record_id_t record_id) override;
   #define TYPE_MACRO(type, return_type, type_id, R, W)\
   void update_##type_id(table_id_t table_id,\
                         record_id_t record_id,\
                         field_id_t field_id,\
                         return_type value) override;\
   void update_vector_##type_id(table_id_t table_id,\
                                record_id_t record_id,\
                                field_id_t field_id,\
                                record_id_t size,\
                                const type *value) override;
   #include "TYPE_MACRO.h"
   #undef TYPE_MACRO
 };
}

#endif
