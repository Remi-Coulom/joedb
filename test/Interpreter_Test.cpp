#include "joedb/ui/Interpreter.h"
#include "joedb/ui/Interpreter_Dump_Writable.h"
#include "joedb/ui/SQL_Dump_Writable.h"
#include "joedb/ui/Raw_Dump_Writable.h"
#include "joedb/interpreted/Database.h"
#include "joedb/journal/Interpreted_File.h"
#include "joedb/journal/Readonly_Journal.h"
#include "joedb/journal/Writable_Journal.h"
#include "joedb/Multiplexer.h"
#include "gtest/gtest.h"

#include <fstream>
#include <sstream>

namespace joedb::ui
{
 /////////////////////////////////////////////////////////////////////////////
 TEST(Interpreter_Test, main_test)
 /////////////////////////////////////////////////////////////////////////////
 {
  Memory_File file;

  std::remove("bio.txt.tmp");

  {
   Writable_Journal journal(file);
   interpreted::Database db;
   Multiplexer multiplexer{db, journal};
   Interpreter interpreter(db, multiplexer, &journal, multiplexer, 0);

   std::ifstream in_file("interpreter_test.joedbi");
   ASSERT_TRUE(in_file.good());
   std::ostringstream out_string;
   interpreter.main_loop(in_file, out_string);
   std::ofstream("interpreter_test.out.tmp") << out_string.str();

   std::ifstream reference_file("interpreter_test.out");
   ASSERT_TRUE(reference_file.good());
   std::ostringstream reference_string;
   reference_string << reference_file.rdbuf();

   EXPECT_EQ(reference_string.str(), out_string.str());
  }
 }

 /////////////////////////////////////////////////////////////////////////////
 TEST(Interpreter_Test, Interpreter_Dump_Writable)
 /////////////////////////////////////////////////////////////////////////////
 {
  interpreted::Database db;
  Memory_File file;
  Writable_Journal journal(file);
  std::ostringstream dump_string;
  const bool blob_wanted = true;
  Interpreter_Dump_Writable writable(dump_string, blob_wanted);
  Multiplexer multiplexer{db, journal, writable};
  Interpreter interpreter(db, multiplexer, &journal, multiplexer, 0);

  std::ifstream in_file("interpreter_test.joedbi");
  ASSERT_TRUE(in_file.good());
  std::ostringstream out_string;
  interpreter.main_loop(in_file, out_string);
  std::ofstream("interpreter_test.dump.tmp") << dump_string.str();

  std::ifstream reference_file("interpreter_test.dump");
  ASSERT_TRUE(reference_file.good());
  std::ostringstream reference_string;
  reference_string << reference_file.rdbuf();

  EXPECT_EQ(reference_string.str(), dump_string.str());
 }

 /////////////////////////////////////////////////////////////////////////////
 TEST(Interpreter_Test, SQL_Dump_Writable)
 /////////////////////////////////////////////////////////////////////////////
 {
  interpreted::Database db;
  Memory_File file;
  Writable_Journal journal(file);
  std::ostringstream dump_string;
  SQL_Dump_Writable writable(dump_string);
  writable.on_blob(Blob(), file);
  Multiplexer multiplexer{db, journal, writable};
  Interpreter interpreter(db, multiplexer, &journal, multiplexer, 0);

  std::ifstream in_file("interpreter_test.joedbi");
  ASSERT_TRUE(in_file.good());
  std::ostringstream out_string;
  interpreter.main_loop(in_file, out_string);
  std::ofstream("interpreter_test.sql.tmp") << dump_string.str();

  std::ifstream reference_file("interpreter_test.sql");
  ASSERT_TRUE(reference_file.good());
  std::ostringstream reference_string;
  reference_string << reference_file.rdbuf();

  EXPECT_EQ(reference_string.str(), dump_string.str());
 }

 /////////////////////////////////////////////////////////////////////////////
 TEST(Interpreter_Test, Raw_Dump_Writable)
 /////////////////////////////////////////////////////////////////////////////
 {
  interpreted::Database db;
  std::ostringstream dump_string;
  Raw_Dump_Writable writable(dump_string);
  Multiplexer multiplexer{db, writable};
  Interpreter interpreter(db, multiplexer, nullptr, multiplexer, 0);

  std::ifstream in_file("interpreter_test.joedbi");
  ASSERT_TRUE(in_file.good());
  std::ostringstream out_string;
  interpreter.main_loop(in_file, out_string);
  std::ofstream("interpreter_test.raw.tmp") << dump_string.str();

  std::ifstream reference_file("interpreter_test.raw");
  ASSERT_TRUE(reference_file.good());
  std::ostringstream reference_string;
  reference_string << reference_file.rdbuf();

  EXPECT_EQ(reference_string.str(), dump_string.str());
 }

 /////////////////////////////////////////////////////////////////////////////
 TEST(Interpreter, Readonly_Interpreted_File)
 /////////////////////////////////////////////////////////////////////////////
 {
  std::stringstream ss;
  joedb::Readonly_Interpreted_File file(ss);
  EXPECT_ANY_THROW(joedb::Writable_Journal{file});
 }

 /////////////////////////////////////////////////////////////////////////////
 TEST(Interpreter, Interpreted_File)
 /////////////////////////////////////////////////////////////////////////////
 {
  std::stringstream ss;
  ss << "create_table person\n";
  ss << "insert_into person 0\n";
  ss << "insert_into person 0\n";
  ss << "insert_into person 0\n";
  ss << "create_table city\n";

  joedb::Readonly_Interpreted_File file(ss);
  joedb::Readonly_Journal journal(file);
  interpreted::Database db;
  journal.play_until_checkpoint(db);
  EXPECT_EQ(db.get_tables().size(), 2ULL);
 }

 /////////////////////////////////////////////////////////////////////////////
 TEST(Interpreter, Writable_Interpreted_File)
 /////////////////////////////////////////////////////////////////////////////
 {
  std::stringstream ss;

  {
   joedb::Interpreted_Stream_File file(ss);
   joedb::Writable_Journal journal(file);
   journal.rewind();
   journal.create_table("person");
   journal.create_table("city");
   journal.default_checkpoint();
   journal.insert_into(Table_Id{1}, Record_Id{1});
   journal.default_checkpoint();
  }

  EXPECT_EQ(ss.str(), "create_table person\ncreate_table city\n\ninsert_into person 1\n");

  joedb::Readonly_Interpreted_File file(ss);
  joedb::Readonly_Journal journal(file);
  interpreted::Database db;
  journal.play_until_checkpoint(db);
  EXPECT_EQ(db.get_tables().size(), 2ULL);
  EXPECT_EQ(db.get_tables().begin()->first, Table_Id{1});
  EXPECT_EQ((++db.get_tables().begin())->first, Table_Id{2});
  EXPECT_EQ(db.get_freedom(Table_Id{1}).size(), 1);
 }
}
