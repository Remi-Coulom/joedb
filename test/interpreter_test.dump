create_table city
create_table person
add_field city name string
add_field person name string
add_field person city references city
add_field person bio blob
insert_into city 1
update city 1 name "Paris"
insert_into city 2
update city 2 name "Lille"
insert_into city 15
update city 15 name "Pétaouchnok"
insert_into person 1
update person 1 name "Rémi"
update person 1 city 2
update person 1 bio 42
insert_into person 2
update person 2 name "Norbert"
update person 2 city 2
update person 2 bio 42
insert_into person 3
update person 3 name "Alexandre"
update person 3 city 15
update person 3 bio 42
drop_table city
drop_table person
create_table test
add_field test a int32
add_field test b int64
insert_into test 1
update test 1 a 1234
update test 1 b 5678
add_field test d string
add_field test e boolean
drop_field test a
insert_into test 2
update test 2 b 3
update test 2 d "4"
update test 2 e 0
delete_from test 2
update test 1 b 56789
create_table ref
add_field ref ref references test
drop_table test
drop_table ref
create_table alltypes
add_field alltypes s string
add_field alltypes i32 int32
add_field alltypes i64 int64
add_field alltypes ref references alltypes
add_field alltypes b boolean
add_field alltypes f float32
add_field alltypes d float64
add_field alltypes o int8
add_field alltypes oulala int16
insert_into alltypes 1
update alltypes 1 s "toto"
update alltypes 1 i32 123456789
update alltypes 1 i64 123456789123456789
update alltypes 1 ref 1
update alltypes 1 b 1
update alltypes 1 f 3.14
update alltypes 1 d 3.14159
update alltypes 1 o 42
update alltypes 1 oulala 12345
timestamp 1234567890 2009-02-13 23:31:30 GMT
comment "This is a comment. これは日本語です。"
valid_data
custom toto_le_petit_rigolo
rename_table alltypes all_types
rename_field all_types s str
create_table float
add_field float value float32
insert_vector float 1 5
update_vector float 1 value 5 0.1 0.2 0.3 0.4 0.5
create_table toto
add_field toto value string
insert_vector toto 3 1
insert_vector toto 5 1
update toto 3 value "toto"
