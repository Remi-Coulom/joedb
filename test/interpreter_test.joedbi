###############################################################################
# Testing basic interpreter commands
###############################################################################
create_table city
create_table person
add_field city name string
add_field person name string
add_field person city references city
insert_into city 1 Paris
insert_into city 2 Lille
insert_into person 1 RÃ©mi 2
insert_into person 2 Norbert 2
insert_into person 3 Alexandre 1
dump

drop_table city
drop_table person
dump

create_table test
add_field test a int32
add_field test a int32
add_field test b int64
add_field test c qsdfqsdf
drop_table qsdfqsdf
qsdfqsdf
insert_into test 1 1234 5678
insert_into test 1 1234 5678
add_field test d string
dump

drop_field test qsdfqsdf
drop_field test a
insert_into test 2 3 4
delete_from test 2
delete_from test 123
update test 1 a 56789
update test 2 b 56789
update test 1 b 56789
dump

create_table ref
add_field ref ref references test
drop_table test
dump

quit
