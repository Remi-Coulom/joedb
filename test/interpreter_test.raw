write_blob "This is my life"
create_table "city"
create_table "person"
add_field 1 "name" string
add_field 2 "name" string
add_field 2 "city" references 1
add_field 2 "bio" blob
insert_into 1 1
update_string 1 1 1 "Paris"
insert_into 1 2
update_string 1 2 1 "Lille"
insert_into 1 15
update_string 1 15 1 "Pétaouchnok"
insert_into 2 1
update_string 2 1 1 "Rémi"
update_reference 2 1 2 2
update_blob 2 1 3 43,15
insert_into 2 2
update_string 2 2 1 "Norbert"
update_reference 2 2 2 2
update_blob 2 2 3 43,15
insert_into 2 3
update_string 2 3 1 "Alexandre"
update_reference 2 3 2 15
update_blob 2 3 3 43,15
drop_table 1
drop_table 2
create_table "test"
add_field 3 "a" int32
add_field 3 "b" int64
insert_into 3 1
update_int32 3 1 1 1234
update_int64 3 1 2 5678
add_field 3 "d" string
add_field 3 "e" boolean
drop_field 3 1
insert_into 3 2
update_int64 3 2 2 3
update_string 3 2 3 "4"
update_boolean 3 2 4 false
delete_from 3 2
update_int64 3 1 2 56789
create_table "ref"
add_field 4 "ref" references 3
drop_table 3
drop_table 4
create_table "alltypes"
add_field 5 "s" string
add_field 5 "i32" int32
add_field 5 "i64" int64
add_field 5 "ref" references 5
add_field 5 "b" boolean
add_field 5 "f" float32
add_field 5 "d" float64
add_field 5 "o" int8
add_field 5 "oulala" int16
insert_into 5 1
update_string 5 1 1 "toto"
update_int32 5 1 2 123456789
update_int64 5 1 3 123456789123456789
update_reference 5 1 4 1
update_boolean 5 1 5 true
update_float32 5 1 6 3.14
update_float64 5 1 7 3.14159
update_int8 5 1 8 42
update_int16 5 1 9 12345
timestamp 1234567890
comment "This is a comment. これは日本語です。"
valid_data
custom "toto_le_petit_rigolo"
rename_table 5 "all_types"
rename_field 5 1 "str"
create_table "float"
add_field 6 "value" float32
insert_vector 6 1 5
update_vector_float32 6 1 1 5 0.1 0.2 0.3 0.4 0.5
create_table "toto"
add_field 7 "value" string
insert_vector 7 3 1
insert_vector 7 5 1
update_string 7 3 1 "toto"
insert_vector 6 6 20
delete_vector 6 11 10
