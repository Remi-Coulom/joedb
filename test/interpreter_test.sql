CREATE TABLE "city"("__id" INTEGER PRIMARY KEY);
CREATE TABLE "person"("__id" INTEGER PRIMARY KEY);
ALTER TABLE "city" ADD "name" TEXT;
ALTER TABLE "person" ADD "name" TEXT;
ALTER TABLE "person" ADD "city" INTEGER REFERENCES "city";
ALTER TABLE "person" ADD "bio" BLOB;
INSERT INTO "city"("__id") VALUES(0);
UPDATE "city" SET "name" = X'5061726973' WHERE "__id" = 0;
INSERT INTO "city"("__id") VALUES(1);
UPDATE "city" SET "name" = X'4c696c6c65' WHERE "__id" = 1;
INSERT INTO "city"("__id") VALUES(14);
UPDATE "city" SET "name" = X'50c3a974616f7563686e6f6b' WHERE "__id" = 14;
INSERT INTO "person"("__id") VALUES(0);
UPDATE "person" SET "name" = X'52c3a96d69' WHERE "__id" = 0;
UPDATE "person" SET "city" = 1 WHERE "__id" = 0;
UPDATE "person" SET "bio" = X'54686973206973206d79206c696665' WHERE "__id" = 0;
INSERT INTO "person"("__id") VALUES(1);
UPDATE "person" SET "name" = X'4e6f7262657274' WHERE "__id" = 1;
UPDATE "person" SET "city" = 1 WHERE "__id" = 1;
UPDATE "person" SET "bio" = X'54686973206973206d79206c696665' WHERE "__id" = 1;
INSERT INTO "person"("__id") VALUES(2);
UPDATE "person" SET "name" = X'416c6578616e647265' WHERE "__id" = 2;
UPDATE "person" SET "city" = 14 WHERE "__id" = 2;
UPDATE "person" SET "bio" = X'54686973206973206d79206c696665' WHERE "__id" = 2;
DROP TABLE "city";
DROP TABLE "person";
CREATE TABLE "test"("__id" INTEGER PRIMARY KEY);
ALTER TABLE "test" ADD "a" INTEGER;
ALTER TABLE "test" ADD "b" BIGINT;
INSERT INTO "test"("__id") VALUES(0);
UPDATE "test" SET "a" = 1234 WHERE "__id" = 0;
UPDATE "test" SET "b" = 5678 WHERE "__id" = 0;
ALTER TABLE "test" ADD "d" TEXT;
ALTER TABLE "test" ADD "e" SMALLINT;
ALTER TABLE "test" DROP COLUMN "a";
INSERT INTO "test"("__id") VALUES(1);
UPDATE "test" SET "b" = 3 WHERE "__id" = 1;
UPDATE "test" SET "d" = X'34' WHERE "__id" = 1;
UPDATE "test" SET "e" = 0 WHERE "__id" = 1;
DELETE FROM "test" WHERE "__id" = 1;
UPDATE "test" SET "b" = 56789 WHERE "__id" = 0;
CREATE TABLE "ref"("__id" INTEGER PRIMARY KEY);
ALTER TABLE "ref" ADD "ref" INTEGER REFERENCES "test";
DROP TABLE "test";
DROP TABLE "ref";
CREATE TABLE "alltypes"("__id" INTEGER PRIMARY KEY);
ALTER TABLE "alltypes" ADD "s" TEXT;
ALTER TABLE "alltypes" ADD "i32" INTEGER;
ALTER TABLE "alltypes" ADD "i64" BIGINT;
ALTER TABLE "alltypes" ADD "ref" INTEGER REFERENCES "alltypes";
ALTER TABLE "alltypes" ADD "b" SMALLINT;
ALTER TABLE "alltypes" ADD "f" REAL;
ALTER TABLE "alltypes" ADD "d" REAL;
ALTER TABLE "alltypes" ADD "o" SMALLINT;
ALTER TABLE "alltypes" ADD "oulala" SMALLINT;
INSERT INTO "alltypes"("__id") VALUES(0);
UPDATE "alltypes" SET "s" = X'746f746f' WHERE "__id" = 0;
UPDATE "alltypes" SET "i32" = 123456789 WHERE "__id" = 0;
UPDATE "alltypes" SET "i64" = 123456789123456789 WHERE "__id" = 0;
UPDATE "alltypes" SET "ref" = 1 WHERE "__id" = 0;
UPDATE "alltypes" SET "b" = 1 WHERE "__id" = 0;
UPDATE "alltypes" SET "f" = 3.14 WHERE "__id" = 0;
UPDATE "alltypes" SET "d" = 3.14159 WHERE "__id" = 0;
UPDATE "alltypes" SET "o" = 42 WHERE "__id" = 0;
UPDATE "alltypes" SET "oulala" = 12345 WHERE "__id" = 0;
-- 2009-02-13 23:31:30 GMT
-- This is a comment. これは日本語です。
-- valid data
-- custom: toto_le_petit_rigolo
ALTER TABLE "alltypes" RENAME TO "all_types";
ALTER TABLE "all_types" RENAME COLUMN "s" TO "str";
CREATE TABLE "float"("__id" INTEGER PRIMARY KEY);
ALTER TABLE "float" ADD "value" REAL;
INSERT INTO "float"("__id") VALUES(0);
INSERT INTO "float"("__id") VALUES(1);
INSERT INTO "float"("__id") VALUES(2);
INSERT INTO "float"("__id") VALUES(3);
INSERT INTO "float"("__id") VALUES(4);
UPDATE "float" SET "value" = 0.1 WHERE "__id" = 0;
UPDATE "float" SET "value" = 0.2 WHERE "__id" = 1;
UPDATE "float" SET "value" = 0.3 WHERE "__id" = 2;
UPDATE "float" SET "value" = 0.4 WHERE "__id" = 3;
UPDATE "float" SET "value" = 0.5 WHERE "__id" = 4;
CREATE TABLE "toto"("__id" INTEGER PRIMARY KEY);
ALTER TABLE "toto" ADD "value" TEXT;
INSERT INTO "toto"("__id") VALUES(2);
INSERT INTO "toto"("__id") VALUES(4);
UPDATE "toto" SET "value" = X'746f746f' WHERE "__id" = 2;
INSERT INTO "float"("__id") VALUES(5);
INSERT INTO "float"("__id") VALUES(6);
INSERT INTO "float"("__id") VALUES(7);
INSERT INTO "float"("__id") VALUES(8);
INSERT INTO "float"("__id") VALUES(9);
INSERT INTO "float"("__id") VALUES(10);
INSERT INTO "float"("__id") VALUES(11);
INSERT INTO "float"("__id") VALUES(12);
INSERT INTO "float"("__id") VALUES(13);
INSERT INTO "float"("__id") VALUES(14);
INSERT INTO "float"("__id") VALUES(15);
INSERT INTO "float"("__id") VALUES(16);
INSERT INTO "float"("__id") VALUES(17);
INSERT INTO "float"("__id") VALUES(18);
INSERT INTO "float"("__id") VALUES(19);
INSERT INTO "float"("__id") VALUES(20);
INSERT INTO "float"("__id") VALUES(21);
INSERT INTO "float"("__id") VALUES(22);
INSERT INTO "float"("__id") VALUES(23);
INSERT INTO "float"("__id") VALUES(24);
DELETE FROM "float" WHERE "__id" = 19;
DELETE FROM "float" WHERE "__id" = 18;
DELETE FROM "float" WHERE "__id" = 17;
DELETE FROM "float" WHERE "__id" = 16;
DELETE FROM "float" WHERE "__id" = 15;
DELETE FROM "float" WHERE "__id" = 14;
DELETE FROM "float" WHERE "__id" = 13;
DELETE FROM "float" WHERE "__id" = 12;
DELETE FROM "float" WHERE "__id" = 11;
DELETE FROM "float" WHERE "__id" = 10;
COMMIT; --checkpoint = 610
COMMIT; --checkpoint = 610
