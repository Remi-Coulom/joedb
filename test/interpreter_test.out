OK: ###############################################################################
OK: # Testing basic interpreter commands
OK: ###############################################################################
OK: create_table city
OK: create_table person
OK: add_field city name string
OK: add_field person name string
OK: add_field person city references city
OK: insert_into city 1 "Paris"
OK: insert_into city 2 "Lille"
OK: insert_into city 15 "Pétaouchnok"
OK: insert_into person 1 "Rémi" 2
OK: insert_into person 2 "Norbert" 2
OK: insert_into person 3 "Alexandre" 15
create_table city
create_table person
add_field city name string
add_field person name string
add_field person city references city
insert_vector city 1 2
insert_vector city 15 1
update city 1 name "Paris"
update city 2 name "Lille"
update city 15 name "Pétaouchnok"
insert_vector person 1 3
update person 1 name "Rémi"
update person 2 name "Norbert"
update person 3 name "Alexandre"
update person 1 city 2
update person 2 city 2
update person 3 city 15
OK: dump
CREATE TABLE "city"("__id" INTEGER PRIMARY KEY);
CREATE TABLE "person"("__id" INTEGER PRIMARY KEY);
ALTER TABLE "city" ADD "name" TEXT;
ALTER TABLE "person" ADD "name" TEXT;
ALTER TABLE "person" ADD "city" INTEGER REFERENCES "city";
INSERT INTO "city"("__id") VALUES(1);
INSERT INTO "city"("__id") VALUES(2);
INSERT INTO "city"("__id") VALUES(15);
UPDATE "city" SET "name" = X'5061726973' WHERE "__id" = 1;
UPDATE "city" SET "name" = X'4c696c6c65' WHERE "__id" = 2;
UPDATE "city" SET "name" = X'50c3a974616f7563686e6f6b' WHERE "__id" = 15;
INSERT INTO "person"("__id") VALUES(1);
INSERT INTO "person"("__id") VALUES(2);
INSERT INTO "person"("__id") VALUES(3);
UPDATE "person" SET "name" = X'52c3a96d69' WHERE "__id" = 1;
UPDATE "person" SET "name" = X'4e6f7262657274' WHERE "__id" = 2;
UPDATE "person" SET "name" = X'416c6578616e647265' WHERE "__id" = 3;
UPDATE "person" SET "city" = 2 WHERE "__id" = 1;
UPDATE "person" SET "city" = 2 WHERE "__id" = 2;
UPDATE "person" SET "city" = 15 WHERE "__id" = 3;
OK: sql
{
 "city":
 {
  "__size": 3,
  "name": ["UGFyaXM=", "TGlsbGU=", "UMOpdGFvdWNobm9r"]
 },
 "person":
 {
  "__size": 3,
  "name": ["UsOpbWk=", "Tm9yYmVydA==", "QWxleGFuZHJl"],
  "city": [1, 1, 2]
 }
}
OK: json 1
------------------
  name        city
------------------
1 "Rémi"         2
2 "Norbert"      2
3 "Alexandre"   15
OK: table person

General commands
~~~~~~~~~~~~~~~~
 about
 help
 quit
 echo on|off

Displaying data
~~~~~~~~~~~~~~~
 table <table_name> [<max_column_width>] [start] [length]
 schema
 dump
 sql
 json [<base64>]

Journal
~~~~~~~
 timestamp [<stamp>] (if no value is given, use current time)
 comment "<comment_string>"
 valid_data
 checkpoint
 blob <data_string>

Data definition
~~~~~~~~~~~~~~~
 create_table <table_name>
 drop_table <table_name>
 rename_table <old_table_name> <new_table_name>
 add_field <table_name> <field_name> <type>
 drop_field <table_name> <field_name>
 rename_field <table_name> <old_field_name> <new_field_name>
 custom <custom_name>

 <type> may be:
  string,
  blob,
  int8, int16, int32, int64,
  float32, float64,
  boolean,
  references <table_name>

Data manipulation
~~~~~~~~~~~~~~~~~
 insert_into <table_name> <record_id>
 insert_vector <table_name> <record_id> <size>
 update <table_name> <record_id> <field_name> <value>
 update_vector <table_name> <record_id> <field_name> <N> <v_1> ... <v_N>
 delete_from <table_name> <record_id>

OK: help
OK: 
Error: create_table: name already used: city
Line 21: create_table city
Error: create_table: invalid identifier: 123456
Line 22: create_table 123456
Error: rename_table: name already used: city
Line 23: rename_table person city
Error: rename_field: name already used: name
Line 24: rename_field person city name
OK: 
OK: drop_table city
OK: drop_table person
OK: dump
OK: 
OK: create_table test
OK: add_field test a int32
Error: add_field: name already used
Line 32: add_field test a int32
OK: add_field test b int64
Error: unknown type
Line 34: add_field test c qsdfqsdf
Error: No such table: qsdfqsdf
Line 35: drop_table qsdfqsdf
Error: Unknown command. For a list of available commands, try "help".
Line 36: qsdfqsdf
OK: insert_into test 1 1234 5678
Error: insert: record_id already in use
Line 38: insert_into test 1 1234 5678
OK: add_field test d string
OK: add_field test e boolean
create_table test
add_field test a int32
add_field test b int64
add_field test d string
add_field test e boolean
insert_vector test 1 1
update test 1 a 1234
update test 1 b 5678
update test 1 d ""
update test 1 e 0
OK: dump
OK: 
Error: drop_field: invalid field_id
Line 43: drop_field test qsdfqsdf
OK: drop_field test a
Error: failed parsing value
Line 45: insert_into test 2 3 "4"
OK: delete_from test 2
Error: delete_record: bad record_id
Line 47: delete_from test 123
Error: bad field
Line 48: update test 1 a 56789
Error: update: invalid record_id
Line 49: update test 2 b 56789
OK: update test 1 b 56789
create_table test
add_field test b int64
add_field test d string
add_field test e boolean
insert_vector test 1 1
update test 1 b 56789
update test 1 d ""
update test 1 e 0
OK: dump
OK: 
OK: create_table ref
OK: add_field ref ref references test
OK: drop_table test
create_table ref
add_field ref ref references __unknown_table__
OK: dump
OK: drop_table ref
OK: 
OK: create_table alltypes
OK: add_field alltypes s string
OK: add_field alltypes i32 int32
OK: add_field alltypes i64 int64
OK: add_field alltypes ref references alltypes
OK: add_field alltypes b boolean
OK: add_field alltypes f float32
OK: add_field alltypes d float64
OK: add_field alltypes o int8
OK: add_field alltypes oulala int16
OK: insert_into alltypes 1 "toto" 123456789 123456789123456789 1 1 3.14 3.14159265359 42 12345
OK: 
OK: timestamp 1234567890
OK: comment "This is a comment. これは日本語です。"
OK: valid_data
OK: 
OK: custom toto_le_petit_rigolo
OK: rename_table alltypes all_types
OK: rename_field all_types s str
OK: 
OK: create_table float
OK: add_field float value float32
OK: insert_vector float 1 5
OK: update_vector float 1 value 5 0.1 0.2 0.3 0.4 0.5
Error: bad field
Line 83: update_vector float 1 qsdf 5 0.1 0.2 0.3 0.4 0.5
OK: 
OK: create_table toto
OK: add_field toto value string
OK: insert_vector toto 3 1
OK: insert_vector toto 5 1
OK: update toto 3 value "toto"
Error: update_vector: invalid record_id range
Line 90: update_vector toto 5 value 2 "kiki" "kuku"
OK: 
Error: rename_table: invalid identifier: __toto__
Line 92: rename_table toto __toto__
Error: No such table: azersdf
Line 93: rename_table azersdf toto
Error: add_field: invalid identifier: __field__
Line 94: add_field toto __field__ string
Error: No such table: azersdf
Line 95: add_field azersdf field string
Error: No such table: azersdf
Line 96: drop_field azersdf field
Error: rename_field: invalid identifier: __value__
Line 97: rename_field toto value __value__
Error: No such table: azersdf
Line 98: rename_field azersdf value valux
Error: rename_field: invalid field_id
Line 99: rename_field toto valux valux
Error: No such table: azersdf
Line 100: insert_into azersdf 1
Error: No such table: azersdf
Line 101: insert_vector azersdf 1 1
Error: No such table: azersdf
Line 102: delete_from azersdf 1
OK: 
create_table all_types
create_table float
create_table toto
add_field all_types str string
add_field all_types i32 int32
add_field all_types i64 int64
add_field all_types ref references all_types
add_field all_types b boolean
add_field all_types f float32
add_field all_types d float64
add_field all_types o int8
add_field all_types oulala int16
add_field float value float32
add_field toto value string
insert_vector all_types 1 1
update all_types 1 str "toto"
update all_types 1 i32 123456789
update all_types 1 i64 123456789123456789
update all_types 1 ref 1
update all_types 1 b 1
update all_types 1 f 3.14
update all_types 1 d 3.14159
update all_types 1 o 42
update all_types 1 oulala 12345
insert_vector float 1 5
update float 1 value 0.1
update float 2 value 0.2
update float 3 value 0.3
update float 4 value 0.4
update float 5 value 0.5
insert_vector toto 3 1
insert_vector toto 5 1
update toto 3 value "toto"
update toto 5 value ""
OK: dump
OK: 
OK: quit
